control-plane/csi-driver/src/bin/node/filesystem_ops.rs

```diff
@@ -0,0 +1,300 @@
+    /// Unmount the filesystem if the filesystem uuid and the provided uuid differ.
+    fn unmount_on_fs_id_diff(
+        &self,
+        device_path: &str,
+        fs_staging_path: &str,
+        volume_uuid: &Uuid,
+    ) -> Result<(), Error>;
+    /// Repair the filesystem with specific filesystem utility.
+    async fn repair(
+        &self,
+        device: &str,
+        staging_path: &str,
+        options: &[String],
+        volume_uuid: &Uuid,
+    ) -> Result<(), Error>;
+    /// Set the filesystem uuid after repair.
+    async fn set_uuid(
+        &self,
+        device: &str,
+        staging_path: &str,
+        options: &[String],
+        volume_uuid: &Uuid,
+    ) -> Result<(), Error>;
+}
+
+#[async_trait]
+impl FileSystemOps for Ext4Fs {
+    async fn create(&self, device: &str) -> Result<(), Error> {
+        let binary = format!("mkfs.{}", "ext4");
+        let output = Command::new(&binary)
+            .arg(device)
+            .output()
+            .await
+            .map_err(|error| format!("failed to execute {binary}: {error}"))?;
+        ack_command_output(output, binary)
+    }
+
+    fn mount_flags(&self, mount_flags: Vec<String>) -> Vec<String> {
+        mount_flags
+    }
+
+    fn unmount_on_fs_id_diff(
+        &self,
+        _device_path: &str,
+        _fs_staging_path: &str,
+        _volume_uuid: &Uuid,
+    ) -> Result<(), Error> {
+        Ok(())
+    }
+
+    async fn repair(
+        &self,
+        device: &str,
+        _staging_path: &str,
+        _options: &[String],
+        _volume_uuid: &Uuid,
+    ) -> Result<(), Error> {
+        let binary = "e2fsck".to_string();
+        let output = Command::new(&binary)
+            .arg("-y")
+            .arg("-f")
+            .arg(device)
+            .output()
+            .await
+            .map_err(|error| format!("failed to execute {binary}: {error}"))?;
+
+        trace!(
+            "Output from {} command: {}, status code: {:?}",
+            binary,
+            String::from_utf8(output.stdout.clone()).unwrap(),
+            output.status.code()
+        );
+
+        // For ext4 fs repair using e2fsck the success condition is when status code is 0 or 1.
+        if !output.status.success() && output.status.code() != Some(1) {
+            return Err(format!(
+                "{} command failed: {}",
+                binary,
+                String::from_utf8(output.stderr).unwrap()
+            ));
+        }
+        Ok(())
+    }
+
+    /// Change the uuid of the ext4 filesystem after repair.
+    async fn set_uuid(
+        &self,
+        device: &str,
+        staging_path: &str,
+        options: &[String],
+        volume_uuid: &Uuid,
+    ) -> Result<(), Error> {
+        if let Ok(probed_uuid) = FileSystem::property(device, "UUID") {
+            if probed_uuid == volume_uuid.to_string() {
+                return Ok(());
+            }
+        }
+        self.repair(device, staging_path, options, volume_uuid)
```

Comment:
We should first try to change it, and only if it fails then repair and change.

Reply:
Maybe we can separate set_uuid, in 2, set_uuid and set_uuid_with_repair? And implement this logic
set_uuid_with_repair:
   if set_uuid.ok return early
   repair
   set_uuid
set_uuid:
    if uuid.same return early
    set_uuid
    match uuid